database/metric_correlations.c

```diff
@@ -0,0 +1,290 @@
+                } else {
+                    pd[j].baseline[c] = cn[j];
+                }
+            }
+        }
+    }
+    rrdr_free(rb);
+    if (!pd)
+        return 0;
+
+    //get highlight
+    RRDR *rh = rrd2rrdr(st, max_points, highlight_after, highlight_before, group_method, group_time, options, NULL, context_param_list);
+    if(!rh) {
+        info("Cannot generate metric correlations output with these parameters on this chart.");
+        freez(pd);
+        return 0;
+    } else {
+        if (rh->d != b_dims) {
+            //TODO handle different dims
+            rrdr_free(rh);
+            freez(pd);
+            return 0;
+        }
+        highlight_points = rrdr_rows(rh);
+        for (c = 0; c != rrdr_rows(rh) ; ++c) {
+            RRDDIM *d;
+            for (j = 0, d = rh->st->dimensions ; d && j < rh->d ; ++j, d = d->next) {
+                calculated_number *cn = &rh->v[ c * rh->d ];
+                    pd[j].highlight[c] = cn[j];
+            }
+        }
+    }
+    rrdr_free(rh);
+
+    for (i = 0; i < b_dims; i++) {
+        fill_nan(&pd[i], highlight_points, baseline_points);
+    }
+
+    for (i = 0; i < b_dims; i++) {
+        run_diffs_and_rev(&pd[i], highlight_points, baseline_points);
+    }
+
+    double d=0, prob=0;
+    for (i=0;i < j ;i++) {
+        if (baseline_points && highlight_points) {
+            kstwo(pd[i].baseline_diffs, baseline_points-1, pd[i].highlight_diffs, highlight_points-1, &d, &prob);
+            buffer_sprintf(wb, "\t\t\t\t\"%s\": %f", pd[i].dimension, prob);
+            if (i != j-1)
+                buffer_sprintf(wb, ",\n");
+            else
+                buffer_sprintf(wb, "\n");
+        }
+    }
+
+    freez(pd);
+    return j;
+}
+
+void metric_correlations (RRDHOST *host, BUFFER *wb, long long baseline_after, long long baseline_before, long long highlight_after, long long highlight_before, long long max_points)
+{
+    info ("Running metric correlations, highlight_after: %lld, highlight_before: %lld, baseline_after: %lld, baseline_before: %lld, max_points: %lld", highlight_after, highlight_before, baseline_after, baseline_before, max_points);
+
+    if (!enable_metric_correlations) {
+        error("Metric correlations functionality is not enabled.");
+        buffer_strcat(wb, "{\"error\": \"Metric correlations functionality is not enabled.\" }");
+        return;
+    }
+
+    if (highlight_before <= highlight_after || baseline_before <= baseline_after) {
+        error("Invalid baseline or highlight ranges.");
+        buffer_strcat(wb, "{\"error\": \"Invalid baseline or highlight ranges.\" }");
+        return;
+    }
+
+    long long dims = 0, total_dims = 0;
+    RRDSET *st;
+    size_t c = 0;
+    BUFFER *wdims = buffer_create(1000);
+
+    if (!max_points || max_points > MAX_POINTS)
+        max_points = MAX_POINTS;
+
+    //dont lock here and wait for results
+    //get the charts and run mc after
+    //should not be a problem for the query
```

Comment:
To do this, we'd also need to set/unset an atomic flag that disabled deletion of charts.

Reply:
Yes! We discussed this already, thanks Vasili. I will look into it.

Reply:
I've added locks with `rrdset_rdlock(st)`, should do it...

Reply:
> I've added locks with rrdset_rdlock(st), should do it... ?

Not sure either... Can you run this under a thread sanitizer and hit the endpoint?

Reply:
> I've added locks with rrdset_rdlock(st), should do it... ?

Not sure either... Can you run this under a thread sanitizer and hit the endpoint?

Reply:
It should be ok... I've tried it, (after reducing number of charts to ~50) but we could discuss it from Monday maybe. We could also add setting `st->last_accessed_time` to keep it from getting obsolete for a while.